scripts/build.py

```diff
@@ -684,26 +823,137 @@ def get_recently_changed_filenames(dev_dir, out_dir):
+            source_dir, staging_dir, mandatory_file_lists, file_hashes)
+        dev_dir_hashes = get_file_hashes(source_dir)
+        print 'Getting files that have changed between %s and %s' % (
+            source_dir, out_dir)
+        recently_changed_filenames = get_recently_changed_filenames(
+            dev_dir_hashes, out_dir)
+        if recently_changed_filenames:
+            print 'Re-building recently changed files at %s' % source_dir
+            build_tasks += generate_build_tasks_to_build_files_from_filepaths(
+                source_dir, staging_dir, recently_changed_filenames,
+                file_hashes)
+            # Clean up files in staging directory that have been removed from
+            # source directory.
+            _execute_tasks(generate_delete_tasks_to_remove_deleted_files(
+                dev_dir_hashes, staging_dir))
+        else:
+            print 'No changes detected. Using previously built files.'
+    return build_tasks
+
+
+def _verify_build(input_dirnames, output_dirnames, file_hashes):
+    """Verify a few metrics after build process finishes:
+        1) Number of files between staging directory and final directory
+        matches.
+        2) The hashes in filenames belongs to the hash dict.
+        3) hashes.js, third_party.min.css and third_party.min.js are built and
+        hashes are inserted.
+
+        Args:
+            input_dirnames: list(str). List of directory paths that contain
+                source files
+            output_dirnames: list(str). List of directory paths that contain
+                built files.
+            file_hashes: dict(str, str). Dictionary with filepaths as keys and
+                hashes of file content as values.
+    """
+    for i in xrange(len(input_dirnames)):
+        # Make sure that all files in source directory and staging directory are
+        # accounted for.
+        _compare_file_count(input_dirnames[i], input_dirnames[i])
+    # Make sure that hashed file name matches with current hash dict.
+    for built_dir in output_dirnames:
+        for root, _, filenames in os.walk(built_dir):
+            for filename in filenames:
+                parent_dir = os.path.basename(root)
+                converted_filepath = os.path.join(
+                    THIRD_PARTY_GENERATED_DEV_DIR, parent_dir, filename)
+                if not hash_should_be_inserted(converted_filepath):
+                    # These filenames should not be hashed.
+                    continue
+                # Obtain the same filepath format as the hash dict's key.
+                filepath = os.path.join(root, filename)
+                relative_filepath = os.path.relpath(filepath, built_dir)
+                _match_filename_with_hashes(relative_filepath, file_hashes)
+
+    hash_final_filename = _insert_hash(
+        HASHES_JS_FILENAME, file_hashes[HASHES_JS_FILENAME])
+
+    MINIFIED_THIRD_PARTY_JS_RELATIVE_FILEPATH = os.path.join(
```

Comment:
It's probably better to pull some of these constants out to top-level global constants and reuse them in the places you use e.g. MINIFIED_THIRD_PARTY_CSS_FILEPATH (you can use os.path.join there). That way, if it's changed in one place, it automatically gets changed in the other place as well. Helps with maintainability.

Reply:
Done!